### DISCUSSION TITLE
Python apps do not start

### DISCUSSION THREAD

**Discussion by andrecolz** [OnionUI/Onion#1566] [Q&A] [answered]
hello everyone, I have created a simple python app for my MMP.
Looking around I saw that it is possible to run python2.7 on MM, then I started to create an app with pygame.
I created the config.json and launch.sh files like all other apps, I included the app in the App folder.
If I start my MMP in the App I see mine with title, icon etc. but if I try to start it it does not work, this because it cannot find python.
Maybe python only runs with "Ports" emulator? is there any way to start it from App too? or should I use another programming language? I saw that all applications are developed in C right?

**Comment by Aemiii91** [comment, marked as answer]
There are some environment variables that need to be set/patched. See here: [Emu/PORTS/launch_python2.sh#L47](https://github.com/OnionUI/Onion/blob/3ad5f32bfe89be7e40e553d675c7c954930e53b3/static/packages/Emu/Ports%20Collection/Emu/PORTS/launch_python2.sh#L47)

**Reply by andrecolz** [reply]
Sorry but I don’t use much python, line 47 should be added to the file. sh right?
Don’t need to edit anything in the python file? 
In case I did so but still not find micropython, I saw a user on reddit that started it with . /micropython but it doesn’t start.

**Reply by Aemiii91** [reply]
Yes, those lines should be in your launch.sh

**Reply by Aemiii91** [reply]
And you launch it using:
```sh
/mnt/SDCARD/.tmp_update/bin/parasyte/python2 ./path_to/file.py
```

**Reply by andrecolz** [reply]
I don’t know what I’m doing wrong.
this is my launch.sh:
```
#!/bin/bash
ParasytePath="/mnt/SDCARD/.tmp_update/lib/parasyte"
export PYTHONPATH=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export PYTHONHOME=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export LD_LIBRARY_PATH=$ParasytePath:$ParasytePath/python2.7/:$ParasytePath/python2.7/lib-dynload:$LD_LIBRARY_PATH
cd "$(dirname "$0")"
/mnt/SDCARD/.tmp_update/bin/parasyte/python2.7 ./main.py
```
the main.py file is in the same directory as launch.sh

**Reply by Aemiii91** [reply]
Looks like you're trying to launch "python2.7" and not "python2"

**Reply by andrecolz** [reply]
even changing the version will not start the same.

**Reply by oherau** [reply]
Some things you could try:
- launch the terminal app and try the python interpreter and check version and if it works properly
- check the execution path by adding the pwd command in your script
- check environnement values using env command in the terminal and/or in your shellscript for debug
- provide the full path to your main.py

**Reply by andrecolz** [reply]
I made some test, i created a .sh file only for install environment variables.
```
#!/bin/bash

ParasytePath="/mnt/SDCARD/.tmp_update/lib/parasyte"
export PYTHONPATH=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export PYTHONHOME=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export LD_LIBRARY_PATH=$ParasytePath:$ParasytePath/python2.7/:$ParasytePath/python2.7/lib-dynload:$LD_LIBRARY_PATH
```
but only LD_LIBRARY_PATH has been set
```
/mnt/SDCARD # echo $PYTHONPATH

/mnt/SDCARD # echo $PYTHONHOME

/mnt/SDCARD # echo $LD_LIBRARY_PATH
/lib:/config/lib:/mnt/SDCARD/miyoo/lib:/mnt/SDCARD/.tmp_update/lib:/mnt/SDCARD/.tmp_update/lib/parasyte
/mnt/SDCARD # python
sh: python: not found
```
So i cant try python interpreter or check version.

**Reply by oherau** [reply]
Settings $PYTHONPATH IS not mandatory unless you have some specific environnement 
You can check this post:
[https://stackoverflow.com/questions/7850908/what-exactly-should-be-set-in-pythonpath](https://stackoverflow.com/questions/7850908/what-exactly-should-be-set-in-pythonpath)

If not already done you could also locate Python and add the containing folder to your path:
`whereis python`
`whereis python2`
Or ugly find-grep:
`find / * | grep python`

**Reply by andrecolz** [reply]
The whereis commands cannot find them, so I used find and this is the output:
```
/mnt/SDCARD # find / * | grep python
/mnt/SDCARD/.tmp_update/bin/parasyte/python
/mnt/SDCARD/.tmp_update/bin/parasyte/python2
/mnt/SDCARD/.tmp_update/bin/parasyte/python2.7
/mnt/SDCARD/.tmp_update/lib/parasyte/python2.7
```
I then modified my file . sh so and I removed PYTHONPATH:
```
#!/bin/bash
ParasytePath="/mnt/SDCARD/.tmp_update/bin/parasyte"
export PYTHONHOME=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export LD_LIBRARY_PATH=$ParasytePath:$ParasytePath/python2.7/:$ParasytePath/python2.7/lib-dynload:$LD_LIBRARY_PATH
```

using the env command I saw that the two environment variables were set but when I write python/python2/python2.7 in the cmd it still can’t find them

**Reply by oherau** [reply]
First things first, you must launch the interpreter properly.
As Aemiii91 said, you must launch the python interpreter with its FULL path:
`/mnt/SDCARD/.tmp_update/bin/parasyte/python2`

The find command retrived it so it is present and you should be able to launch it

Then you can provide your *.py as an argument to the interpreter ALSO with the full absolute path.

Then if you still have issues with the interpreter, you can do some fine tuning with environnement variables

**Reply by andrecolz** [reply]
Thank you all, I managed to fix the problem,  unfortunately I have never developed on linux and I am new to this environment

**Reply by Aemiii91** [reply]
Please state how you fixed the problem for future reference 😊

**Reply by andrecolz** [reply]
I fixed the problem just like you said at the beginning by putting the environment variables in the launch.sh file
the problem is that I started it from the miyoo app terminal or from bash shell (ssh), and in this case I had to run ctrl+c once to view the app, doing it again I closed it.
if I try to start it by clicking on my app from "App" performs "loading" but then nothing happens and returns to the list "App".
```
#!/bin/bash
ParasytePath="/mnt/SDCARD/.tmp_update/lib/parasyte"
export PYTHONPATH=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export PYTHONHOME=$ParasytePath/python2.7:$ParasytePath/python2.7/site-packages:$ParasytePath/python2.7/lib-dynload
export LD_LIBRARY_PATH=$ParasytePath:$ParasytePath/python2.7/:$ParasytePath/python2.7/lib-dynload:$LD_LIBRARY_PATH

/mnt/SDCARD/.tmp_update/bin/parasyte/python2 main.py
```

even putting the entire path of the main.py directory with and without . / or .. / does not work on bash shell that from "App"

**Reply by Aemiii91** [reply]
OK, so you only got it working through the built-in Terminal app? (If I understand you correctly)

**Reply by andrecolz** [reply]
Yes it only works in terminal app (launch.sh is how I sent it above).
It’s probably a bug but in this case every now and then there are flicker effects (the app only shows a green background) and you can see pieces of the terminal app sometimes.

**Reply by Aemiii91** [reply]
That makes sense, since Terminal and  your app are probably fighting over the display.

**Reply by andrecolz** [reply]
Okk, but for start it from "App" is possible or with python is not possible?
When i click on some app the script from what directory start? The directory of selected app or from some script .tmp?
